include/deal.II/numerics/nonlinear.h

```diff
@@ -0,0 +1,582 @@
+     * @p KINSOL documentation.
+     */
+    SolutionStrategy strategy = linesearch;
+
+
+    /**
+     * Maximum number of nonlinear iterations allowed.
+     */
+    unsigned int maximum_non_linear_iterations;
+
+    /**
+     * A scalar used as a stopping tolerance on the scaled
+     * maximum norm of the system function $F(u)$ or $G(u)$.
+     *
+     * If set to zero, default values provided by KINSOL will be used.
+     */
+    double function_tolerance;
+
+    /**
+     * A scalar used as a stopping tolerance on the minimum
+     * scaled step length.
+     *
+     * If set to zero, default values provided by KINSOL will be used.
+     */
+    double step_tolerance = 0.0;
+
+    /**
+     * Whether an initial call to the preconditioner or Jacobian
+     * setup function should be made or not.
+     *
+     * A call to this function is useful when solving a sequence of problems,
+     * in which the final preconditioner or Jacobian value from one problem
+     * is to be used initially for the next problem.
+     */
+    bool no_init_setup = false;
+
+
+    /**
+     * The maximum number of nonlinear iterations that can be
+     * performed between calls to the setup_jacobian() function.
+     *
+     * If set to zero, default values provided by KINSOL will be used,
+     * and in practice this often means that KINSOL will re-use a
+     * Jacobian matrix computed in one iteration for later iterations.
+     */
+    unsigned int maximum_setup_calls = 0;
+
+    /**
+     * The maximum allowable scaled length of the Newton step.
+     *
+     * If set to zero, default values provided by KINSOL will be used.
+     */
+    double maximum_newton_step = 0.0;
+
+    /**
+     * The relative error in computing $F(u)$, which is used in the
+     * difference quotient approximation to the Jacobian matrix when the user
+     * does not supply a solve_jacobian_system_matrix() function.
+     *
+     * If set to zero, default values provided by KINSOL will be used.
+     */
+    double dq_relative_error = 0.0;
+
+    /**
+     * The maximum number of beta-condition failures in the
+     * linesearch algorithm. Only used if
+     * strategy==SolutionStrategy::linesearch.
+     */
+    unsigned int maximum_beta_failures = 0;
+
+    /**
+     * The size of the subspace used with Anderson acceleration
+     * in conjunction with Picard or fixed-point iteration.
+     *
+     * If you set this to 0, no acceleration is used.
+     */
+    unsigned int anderson_subspace_size = 0;
+
+    /**
+     * Relative l2 tolerance of the residual to be reached.
+     *
+     * @note Solver terminates successfully if either the absolute or
+     * the relative tolerance has been reached.
+     */
+    const double relative_tolerance = 1.e-5;
+
+    /**
+     * Number of nonlinear iterations after which the preconditioner
+     * should be updated.
+     */
+    unsigned int threshold_nonlinear_iterations = 0;
```

Comment:
Is there a special meaning to zero? Like "never"? In that case, can you document how zero is interpreted?

Reply:
That parameter can actually be omitted now that I read up on it. That parameter is only useful in NOX if we're using a particular function, which we haven't included in this class. Thank you for pointing this out.